Assets/Mirror/Runtime/QuaternionReadWrite.cs

```diff
@@ -0,0 +1,348 @@
+            uint a2 = byte.MaxValue & (a >> 8);
+
+            uint b1 = byte.MaxValue & b;
+            uint b2 = byte.MaxValue & (b >> 8);
+
+            uint c1 = byte.MaxValue & c;
+            uint c2 = byte.MaxValue & (c >> 8);
+
+            uint extra = ((uint)largestIndex) + (a2 << 2) + (b2 << 3) + (c2 << 4);
+
+            writer.WriteByte((byte)a1);
+            writer.WriteByte((byte)b1);
+            writer.WriteByte((byte)c1);
+            writer.WriteByte((byte)extra);
+        }
+        static Quaternion ReadQuaternionHalf(NetworkReader reader)
+        {
+            Quaternion result;
+            uint a1 = reader.ReadByte();
+            uint b1 = reader.ReadByte();
+            uint c1 = reader.ReadByte();
+            byte extra = reader.ReadByte();
+
+            // first 2 bytes
+            uint largestIndex = (uint)(extra & 3);
+
+            // get nth bit, then move to start
+            uint a2 = (uint)((extra & (1 << 2)) >> 2);
+            uint b2 = (uint)((extra & (1 << 3)) >> 3);
+            uint c2 = (uint)((extra & (1 << 4)) >> 4);
+
+            uint a = a1 | (a2 << 8);
+            uint b = b1 | (b2 << 8);
+            uint c = c1 | (c2 << 8);
+
+            const int bitLength = 9;
+
+            float x = scaleFromUInt(a, bitLength);
+            float y = scaleFromUInt(b, bitLength);
+            float z = scaleFromUInt(c, bitLength);
+
+            Vector3 small = new Vector3(x, y, z);
+            result = FromSmallerDimensions(largestIndex, small);
+            return result;
+        }
+
+
+        internal static int findLargestIndex(Quaternion q)
+        {
+            int index = 0;
+            float current = Mathf.Abs(q.x);
+
+            for (int i = 1; i < 4; i++)
+            {
+                float next = Mathf.Abs(q[i]);
+                if (next > current)
+                {
+                    index = i;
+                    current = next;
+                }
+            }
+
+            return index;
+        }
+
+
+        static Vector3 getSmallerDimensions(int largestIndex, Quaternion value)
+        {
+            float x = value.x;
+            float y = value.y;
+            float z = value.z;
+            float w = value.w;
+
+            float a, b, c;
+            switch (largestIndex)
+            {
+                case 0:
+                    a = y;
+                    b = z;
+                    c = w;
+                    break;
+                case 1:
+                    a = x;
+                    b = z;
+                    c = w;
+                    break;
+                case 2:
+                    a = x;
+                    b = y;
+                    c = w;
+                    break;
+                case 3:
+                    a = x;
+                    b = y;
+                    c = z;
+                    break;
+                default:
+                    throw new System.ArgumentException("LargestIndex did not have value between 0 and 3");
```

Comment:
how likely is this to occur?
we need to be really careful with exceptions. if a user finds out how to trigger an exception on the server that kills the rest of the Update code, then that's basically an exploit

Reply:
Looking at it again I don't think this should ever happen.

I can replace `Exception` and make it return `Quaternion.identity` in this case instead

Reply:
@James-Frowen I see. If it's never supposed to happen then the exception is okay.
Just in generaly it's good if our reader/writer functions simply read/write whatever we throw at them :)

Reply:
I've replaced the exception just to be safe